merge "show direction" and routing modes (fix #9868)

baiti opened the pull request:
improved logging in the context of working issue_9868

moving-bits commented:
@baiti 
Is this for testing purposes only or meant to be included in c:geo? If the latter, then please remove the line you commented out. Thanks.

baiti commented:
I just noticed I can push multiple iterations into an existing and open PR. So pls do not merge yet. Let's wait until we're done with issue_9868.

Lineflyer commented:
@baiti 
I converted the PR into a draft to signal, that you are still working on it. Feel free to convert it back once you consider it ready.

When finished you should squash all commits into a single one for better overview (can be done by some tools or plainly using `git rebase -i`)

baiti commented:
> @baiti
> I converted the PR into a draft to signal, that you are still working on it. Feel free to convert it back once you consider it ready.
> 
> When finished you should squash all commits into a single one for better overview (can be done by some tools or plainly using `git rebase -i`)

This github has nice features ;) I assume I can still push changes into the PR draft? We'll find out soon.

I go and try to eliminate the unnecessary setting from preferences.xml later today. Also I will go and try to find some improvement at the point where BRouter does not return an XML but an error message instead. This happens if the user hasn't loaded the respective map into BRouter. Question is, would we want to display a toast and inform the user about it?

fm-sys commented:
> This happens if the user hasn't loaded the respective map into BRouter. Question is, would we want to display a toast and inform the user about it?

That would be kind of cool! IIRC we also considered in another issue making the download of brouter map files possible directly via c:geo if they are missing...

baiti commented:
> @baiti
> I converted the PR into a draft to signal, that you are still working on it. Feel free to convert it back once you consider it ready.
> 
> When finished you should squash all commits into a single one for better overview (can be done by some tools or plainly using `git rebase -i`)

I'd say I have all my changes done, tested on the emulator, commited, and pushed into the branch of my cloned fork.

At the top of this PR I saw: "Commits: 9"

So I figured I would do a: `git rebase -i HEAD~9` .. which wasn't a great idea because it wanted to pick stuff I clearly did not push:

`pick 051714a6e rel to #9904: better migration handling of legacy values`
`pick c3095c5c8 fix #9903: handle selection of Download folder`
`pick cd00b2f4a Integrate SAF base folder query into wizard`
`pick 54d49899f check db compatibility against backup db on restore (fix #9863)`
`pick c6addea05 log in ParseGpxTrack refined`
`pick dca6595bf issue_9868, removed commented out code line`
`pick 7a552f4d0 log in ParseGpxTrack refined`
`pick 63999e2e9 log in ParseGpxTrack refined`
`pick 3bbad08cf issue_9868, removed commented out code line`
`pick 0390ba45c Routing/Settings changed to reflect #9868`
`pick 0d932a7c9 handle error msg from BRouter`
 

I manually removed those commits such that only the ones I did were listed, it then looked like this:

`pick c6addea05 log in ParseGpxTrack refined`
`pick dca6595bf issue_9868, removed commented out code line`
`pick 7a552f4d0 log in ParseGpxTrack refined`
`pick 63999e2e9 log in ParseGpxTrack refined`
`pick 3bbad08cf issue_9868, removed commented out code line`
`pick 0390ba45c Routing/Settings changed to reflect #9868`
`pick 0d932a7c9 handle error msg from BRouter`

This gave me:

`Auto-merging main/src/cgeo/geocaching/maps/routing/Routing.java`
`CONFLICT (content): Merge conflict in main/src/cgeo/geocaching/maps/routing/Routing.java`
`error: could not apply 7a552f4d0... log in ParseGpxTrack refined`
`Resolve all conflicts manually, mark them as resolved with`
`"git add/rm <conflicted_files>", then run "git rebase --continue".`
`You can instead skip this commit: run "git rebase --skip".`
`To abort and get back to the state before "git rebase", run "git rebase --abort".`
`Could not apply 7a552f4d0... log in ParseGpxTrack refined`

In order not to screw up things I decided to do a `git rebase --abort` and see what the experts have to say ;)

fm-sys commented on main/src/cgeo/geocaching/maps/routing/Routing.java:
```diff
@@ -207,6 +207,14 @@ public static synchronized void disconnect(final String callbackKey) {
+        // wanted to compare (gpx.substr(0,5) != "<?xml")
+        // but this did not work, most likely because gpx is a Unicode string?
+
+        if (gpx.length() < 200) {
```
I think comparing the length isn't a good idea. what about `!gpx.startsWith("<?xml")`?

baiti replied:
Yes, I agree, it is not a good idea, however, if this: `(gpx.substr(0,5) != "<?xml")` did not work, and it didn't, why would that: `!gpx.startsWith("<?xml")` work any better? ... Well ... I will just go ahead and try it.

fm-sys replied:
Why it is not working exactly? Have you tried to debug what is returned by `gpx.substr(0,5)` or is it raising any error? I don't really understand why it shouldn't work...

baiti replied:
I had put an unconditional: `Log.w(gpx.substr(0,5))` in front of the if () statement and as expected, it logged `<?xml` but the if () expression never evaluaet to `true`. I played around with substr(0,4) vs. substr(0,5) but nothing ever made it a `true` so as an act of some sort of desparation I coded the .lenth() thing and it worked right away. I tested both paths, one that returned a valid XML and one that returned an error message.

baiti replied:
I now tried: `!gpx.startsWith("<?xml")` ... and it worked. Perhaps I should go back and try my initial `(gpx.substr(0,5) != "<?xml")` for completeness.....

baiti replied:
Again, didn't work! Why? Beats me. I use your recommendation gpx.startsWith and will push this now.

moving-bits replied:
I guess this is due to the way Java compares strings: If you use "==" comparison operator on strings, only the object addresses will be compared, not the content. To compare the strings' contents, you need method like `string1.equals(string2)` or the helper methods from `StringUtil` like `StringUtil.isNotBlank(string1)` etc.

baiti replied:
Thank you very  much. This explanation resonates well. As you can clearly see, I am new to Java. Something like this would perfectly work in JavaScript.

On the side: I am actually surprised that Google hasn't made the move to put all Android Apps on top of their V8 engine. It would signifiantly simplify the ecosystem as this is what runs in the Browsers as well as in serverside Node.JS which Google is a big fan of for both.

fm-sys reviewed (commented):
I added one comment to your changes...

> In order not to screw up things I decided to do a `git rebase --abort` and see what the experts have to say ;)

IMHO squashing the commits by yourself is not necessary, although it would be nice to have. Something must have gone very wrong there (I also despair in such situations with git). We can simply squash the commits while merging within GitHub ("Squash and Merge"). That should be fine...

bekuno commented:
Can you create a test case with your test data?

baiti commented:
> Can you create a test case with your test data?

My test case was runing it on the Emulator. I have two map areas, (a) one where I have the map tiles loaded in BRouter, an (b) one where I do not have map files loaded in BRouter. Then  I used a location spoofer and placed myself to (a) and verified that I could get a valid GPX back from BRouter, then I swichted locatoin to (b) and did the same and verified that I cot the error message referenced in the string compare discussion.

I would not know how I could create a formal testase (I guess that is what you're aksing?) such that this test could be perfomed automatically and repeatedly by CI.

Lineflyer commented:
> In order not to screw up things I decided to do a git rebase --abort and see what the experts have to say ;)

I would say it was nearly correct, but you have to use "pick" only for one of the commits (ideally the "main" commit) and replace "pick" by squash for all the others you want git to incorporate into the picked commit. This should then result in one (picked) commits containing the other (squashed commits).
BTW: All of this needs to be done while you are on your branch "issue_9868" (if you have been on another branch that probably explains the unrelated commits in your list).

But as @fm-sys mentioned, If you do not manage to squash them, we can do it while merging. Would only be a nice to have for easy overview in the PR.

baiti commented:
> > In order not to screw up things I decided to do a git rebase --abort and see what the experts have to say ;)
> 
> I would say it was nearly correct, but you have to use "pick" only for one of the commits (ideally the "main" commit) and replace "pick" by squash for all the others you want git to incorporate into the picked commit. This should then result in one (picked) commit containing also the other (squashed) commits.
> BTW: All of this needs to be done while you are on your branch "issue_9868" (if you have been on another branch that probably explains the unrelated commits in your list).
> 
> But as @fm-sys mentioned, If you do not manage to squash them, we can do it while merging. Would only be a nice to have for easy overview in the PR.

Oh wow ... I guess I need to find some serious tutorial somewhere on the web. At this moment, I feel like the junior guy in this clip. Does anybody want to sign up for the senior? ;)

https://www.youtube.com/watch?v=rR4n-0KYeKQ

baiti commented:
> > I would say it was nearly correct, but you have to use "pick" only for one of the commits (ideally the "main" commit) and replace "pick" by squash for all the others you want git to incorporate into the picked commit. This should then result in one (picked) commit containing also the other (squashed) commits.

If I did he pick and squash correctly, what would be the result? I would assume that it only changes states on my local working copy of the cloned fork of the repo. Nothing would change with respect to the pull request on github, correct? There are most likely additional commands required, either on my local system or on the github UI, to squash what's in the PR?

moving-bits commented:
> If I did he pick and squash correctly, what would be the result? I would assume that it only changes states on my local working copy of the cloned fork of the repo. Nothing would change with respect to the pull request on github, correct? There are most likely additional commands required, either on my local system or on the github UI, to squash what's in the PR?

Without extra parameters specific to remote repos, all commands like `git rebase`, `git commit` etc. done on your local command line only affect your local repo. Until you hit the `git push` command, which pushes your local changes to your fork. When you have done a `push` already, and do some rebasing or squashing after that, you won't be able to `push` again directly, as those commands change the commit history. You will have to do a so called "forced push" instead, meaning `push --force`.

Personally I use a mixture of shell scripts / batch files and commands entered manually. And some things (like fetching history) I even do from within Android Studio. Additionally, I configured git to use diff tool, so that a `git difftool` shows me a nice visual view of the changes between last commit and current workspace.

fm-sys commented:
When we're on the story:
I personally use android studio's GUI for most Git operations. But that's just a habitual thing. Only some advanced stuff like retargeting still needs to be done at the command line... @baiti In the beginning, I also felt like the junior guy in the posted clip. But you get used to Git quickly until and at some point you don't even think about it anymore.

baiti commented:
So considering all this, I would conclusively say, this PR is now finally ready for review and hopefully the merge. Thanks much to the team for providing support and encouragement.

moving-bits commented:
Aside from the review: This PR cannot be merged currently, as Github sees some file conflicts, see the dark grey triangle with the "This branch cannot be rebased due to conflicts" notice.

To resolve them you need to rebase your PR locally (probably after squashing it locally), which will show you some merge conflicts. Resolve them, do a `git add <filename>` for the conflicting files, and a `git rebase continue`. You may have to do this more than once, depending on the number of conflicting merges in between.

baiti commented:
> Aside from the review: This PR cannot be merged currently, as Github sees some file conflicts, see the dark grey triangle with the "This branch cannot be rebased due to conflicts" notice.
> 
> To resolve them you need to rebase your PR locally (probably after squashing it locally), which will show you some merge conflicts. Resolve them, do a `git add <filename>` for the conflicting files, and a `git rebase continue`. You may have to do this more than once, depending on the number of conflicting merges in between.

I am afraid this is over my head at this time. Here is what I would suggest: I have modified a very small set of files. I will stash those away. We disccard the PR, I throw away my clone, delete my fork .. and start over completely from scratch. I would make sure I do ONE commit / push / create PR then it should work. Does this make sense?

moving-bits commented:
> Does this make sense?

Yes and no :-)

While it might save you from the hassle of resolving the merge conflicts this time, you'll need to establish a ways to keep your local copy up to date anyway.
That's one of the parts I'm using my scripts for. At the core it's something like this:

```
echo updating repo data from upstream
git fetch upstream
echo checking out and updating master branch
git checkout master
git pull --rebase upstream master
git status
```

In this config `origin` links to your fork on github.com, and `upstream` to the original c:geo repo. So what I'm doing here every day when I start working on c:geo (and sometimes in between as well) is updating my local repo to the same level as the original c:geo repo has. My fork of c:geo is way outdated (some 2.000 commits behind, meanwhile) and only used for storing PR before merging them into c:geo.

Doing this regularly reduces the danger of merge conflicts or at least reduces the amount of conflicts.
Also, when I'm working on some larger PR ("large" in terms of lines of code or in terms of timespan between starting to work on it and preparing a PR), I'm regularly doing a `git rebase upstream/master` within that branch, which also reduces the potential for merge conflicts.

fm-sys commented:
> Aside from the review: This PR cannot be merged currently, as Github sees some file conflicts, see the dark grey triangle with the "This branch cannot be rebased due to conflicts" notice.

The PR _can_ be merged when choosing "Squash and merge". Then it shows "merge can be performed automatically"

baiti commented:
> > Does this make sense?
> 
> Yes and no :-)
> 
> While it might save you from the hassle of resolving the merge conflicts this time, you'll need to establish a ways to keep your local copy up to date anyway.

Yes, I understand that thought.

> That's one of the parts I'm using my scripts for. At the core it's something like this:
> 
> ```
> echo updating repo data from upstream
> git fetch upstream
> echo checking out and updating master branch
> git checkout master
> git pull --rebase upstream master
> git status
> ```

I have actually done this rebase sequence multiple times over the last 3 days and I have no idea why we are in the state we are now and how I could potentially get out of this. 

I could execute that sequence again and again, it would not change anything in the PR that you can see and review, and eventually merge. Correct? None of the commands above make any change to the PR on github, I believe.

I have no idea how I can make any changes to the PR's current state and content. I can commit / push more stuff but that will make things only worse.

moving-bits commented:
> > Aside from the review: This PR cannot be merged currently, as Github sees some file conflicts, see the dark grey triangle with the "This branch cannot be rebased due to conflicts" notice.
> 
> The PR _can_ be merged when choosing "Squash and merge". Then it shows "merge can be performed automatically"

Ah, even better! Hadn't checked the squashing variant, sorry. That'll save @baiti quite some trouble :-)

baiti commented:
> > > Aside from the review: This PR cannot be merged currently, as Github sees some file conflicts, see the dark grey triangle with the "This branch cannot be rebased due to conflicts" notice.
> > 
> > 
> > The PR _can_ be merged when choosing "Squash and merge". Then it shows "merge can be performed automatically"
> 
> Ah, even better! Hadn't checked the squashing variant, sorry. That'll save @baiti quite some trouble :-)

Yes, what a **relief** ... but I am fully aware of and without any doubts, that I need to significantly work on my understanding of the process. Thank you very much for your help and your patience

moving-bits commented:
> I have actually done this rebase sequence multiple times over the last 3 days and I have no idea why we are in the state we are now and how I could potentially get out of this.
> I could execute that sequence again and again, it would not change anything in the PR that you can see and review, and eventually merge. Correct? None of the commands above make any change to the PR on github, I believe.

The rebase statements in my last post were only about you local repo's state. As long as you do not push your PR branch back to the server, nothing will change here on GitHub.
And in this special case, squashing seems to be necessary to spare you the merge conflicts. Seems that you have changed a line which conflicts with a change in some other PR, but reverted that change later, so that a squashed PR does not have that conflicting line.

> I have no idea how I can make any changes to the PR's current state and content. I can commit / push more stuff but that will make things only worse.

I'll do a squash & merge here server-side, that's the easiest now.

The first handful of PR are the hardest in terms of getting acquainted with git, but then it will start getting better :-) Thanks for your endurance!

Lineflyer commented:
> The PR can be merged when choosing "Squash and merge". Then it shows "merge can be performed automatically"

That is totally strange as it seems to imply, that the commits inside the PR are conflicting each other...I wonder how this is possible...

baiti commented:
> > The PR can be merged when choosing "Squash and merge". Then it shows "merge can be performed automatically"
> 
> That is totally strange as it seems to imply, that the commits inside the PR are conflicting each other...I wonder how this is possible...

This might be because I believe I have commited / pushed the same things twice. That was at the time I removed that single Log.x() statement that I had put into a comment line before I changed it to what it is now.

baiti commented:
If I now press "Delete branch", which is what github offers me to do .. what do I now have to do on my local system to get everything synced up. Do I also delete the branch on my local system, switch back to master and do some sort of rebase?

moving-bits commented:
> That is totally strange as it seems to imply, that the commits inside the PR are conflicting each other...

Not necessarily. If you change a line in commit #1 which conflicts with some other PR already online, you'll get a merge conflict. If you revert that change with commit #2, a squashed PR including both changes will not conflict any longer, though commit #1 separately still will. At least that's how I'm trying to explain it to myself.

Lineflyer commented:
@

moving-bits commented:
> If I now press "Delete branch", which is what github offers me to do .. what do I now have to do on my local system to get everything synced up. Do I also delete the branch on my local system, switch back to master and do some sort of rebase?

When somebody else merges my PR, I have to delete my PR branch in both my fork and my local copy. Which I have yet another script for :-)
```
if "%1" == "" goto ende
echo Delete branch "%1" (local + remote)...
pause
echo (remote: %1)
git push -d origin %1
echo (local: %1)
git branch -D %1
:ende
```

> Should I delete my fork? In which case I better also delete my cloned copy of that fork on my local system

No, you don't have to delete your fork on github.com. Just update your local copy as described in https://github.com/cgeo/cgeo/pull/9913#issuecomment-771099500.

Lineflyer commented:
@moving-bits Sound indeed logical...need to try it next days to get more practice in conflict resolution

@baiti 
To update your master it should be sufficient to use `git pull upstream master` given that your fork is `origin` and you defined another remote called `upstream` which refers to the real c:geo repo.

baiti commented:
Done all of what you guys described. Deleted the branch remote and local and  pulled and rebased. The pull --rebase ... wow there was a lot going on in these 3 days.

I guess I need to also create some of these "mother's little helper" scripts.

...Now would be a good opportunity to play around with the ALC stuff. Except that we should somewhere have a discussion on how we would want to integrate it. Since ALCs are referenced by some sort of a UUID as opposed to a "GC-code", we could treat it like a completely seprate "platform" and write a connector for it. We would only need two types of API Calls. One to get an icon on the map and one that shows details upon cklicking of such an icon.

At this time, the required two api calls are open and free and don't require ANY credentials. The question is: will that remain ? Or will they lockdown that API?

moving-bits commented:
> ...Now would be a good opportunity to play around with the ALC stuff. Except that we should somewhere have a discussion on how we would want to integrate it.

Please let's continue THAT discussion in #9202. I've copied your last comment over there.

baiti commented:
It seems I have introduced a bug with this PR. Normally I use "walking" mode exclusively for routing but yesterday I switched to "car" and nothing changed. It continued to show a walking route. This is on V 2021.02.14.

I will try to see if I can reproduce / debug this on the emulator later today.

Has this worked correctly (and been tested) before I touched it?

rsudev commented:
I must admit I never used anything else as walking mode either, so I can't tell. @Lineflyer ?

moving-bits commented:
Cannot reproduce here, tested with both OSM offline maps and Google maps:
As soon as I close the "map quick settings" popup after changing from e. g. "walk" to "car" the route gets recalculated according to the chosen method.

Lineflyer commented:
> As soon as I close the "map quick settings" popup after changing from e. g. "walk" to "car" the route gets recalculated according to the chosen method.

Same for me, looks working fine.

fm-sys commented:
I've made two bike tours in the last days and noticed, that sometimes brouter doesn't calculate the route at all (only a straight line) after removing a cache from the individual route (set as target checked). I don't know if this is related to your observation, but anyway it's probably not caused by your changes. Don't worry ;-) 
For me the issue could be solved by closing and reopening c:geo completely...

baiti commented:
> I've made two bike tours in the last days and noticed, that sometimes brouter doesn't calculate the route at all (only a straight line) after removing a cache from the individual route (set as target checked). I don't know if this is related to your observation, but anyway it's probably not caused by your changes. Don't worry ;-)
> For me the issue could be solved by closing and reopening c:geo completely...

Interesting. I have not been able to display any other routes than "walk". If I switch to "car" it still forces me through paths which are clearly not suitable for anything else other than per pedes.

Will investigate that further. This is where the location spoofer now comes into play.
